Make it standard

bettio opened the pull request.

bettio committed:
JSONPath results are always arrays (lists).

Result must be always an array also for member access to nested objects.
Also $.data[0].values[?(@.name == "test")].value[0] is not a valid path,
hence it cannot be used to just return the item itself.

bettio committed:
Return {:ok, result} on success

rbino commented on lib/ex_json_path.ex:
```diff
@@ -24,7 +24,15 @@ defmodule ExJsonPath do
 
   def eval(map, compiled_path) when is_map(map) and is_list(compiled_path) do
-    recurse(map, compiled_path)
+    result = recurse(map, compiled_path)
```
There's one case where you return `{:error, :no_match}` in `recurse`, so this should probably go in a `with` block

bettio committed:
Ensure a consistent behavior

Make sure that results are always consistent with existing
implementations.

bettio committed:
mix.exs: increase version number to 0.2.0

Behavior is going to change, therefore update version number.

bettio committed:
tests: improve descriptions

Make tests descriptions clearer.

bettio committed:
Add match-all function head

Return error tuple by default.

bettio committed:
tests: unmatchable paths must be handled

unmatchable paths must return {:error, :no_match}.

bettio committed:
Allow arrays (lists) as root items

Handle paths such as $[0].

bettio committed:
tests: add test for $[0]

Test $[0] path evaluation against an array.

rbino commented on lib/ex_json_path.ex:
```diff
@@ -17,14 +17,16 @@
 
 defmodule ExJsonPath do
-  def eval(map, path) when is_map(map) and is_binary(path) do
+  defguard is_valid_input(input) when is_map(input) or is_list(input)
```
I think I recall that also bare JSON values are valid JSON (e.g. `3` is a valid JSON document).
In fact, if you put [here](https://jsonpath.com/) `"foo"` as JSON and `$` as JSON path query, it will return `["foo"]` as result.

I think this can be handled in a special function head since the only possible query on raw values is `$`, but I might be wrong.

bettio committed:
Do not crash on bare values

eval on a bare value never matches but doesn't crash.

bettio committed:
tests: add tests for bare values

Make sure that bare values always return {:error, :no_match}.

bettio committed:
tests: make sure that $.a.b on {"a": {"b": {}}} returns [{}]

Most implementations return [{}] which makes sense, while one of the
Goessner's implementations return [[]] which doesn't really makes sense.
